modules/core/src/test/java/org/apache/ignite/internal/processors/service/GridServiceProxyTopologyInitializationTest.java

```diff
@@ -0,0 +1,241 @@
+import static org.apache.ignite.testframework.GridTestUtils.assertThrowsWithCause;
+import static org.apache.ignite.testframework.GridTestUtils.runAsync;
+
+/**
+ * Tests service invocation if requested service is not deployed on the local node and the service topology is required.
+ */
+public class GridServiceProxyTopologyInitializationTest extends GridCommonAbstractTest {
+    /** Name of the service that throws exception during initialization. */
+    private static final String BROKEN_SRVC = "broken-service";
+
+    /** Name of the decent service. */
+    private static final String DECENT_SRVC = "decent-service";
+
+    /** Name of the attribute that shows whether test servise deployment will be skipped on the node. */
+    private static final String ATTR_SKIP_DEPLOYMENT = "skip-deployment";
+
+    /** Latch that indicates whether {@link ServiceSingleNodeDeploymentResultBatch} execution should be proceeded. */
+    private final CountDownLatch fullMsgUnblockedLatch = new CountDownLatch(1);
+
+    /** Latch that indicated whether {@link ServiceSingleNodeDeploymentResultBatch} was received on the remote node. */
+    private final CountDownLatch fullMsgReceivedLatch = new CountDownLatch(1);
+
+    /** Latch that indicated whether {@link ServiceSingleNodeDeploymentResultBatch} was handled on the remote node. */
+    private final CountDownLatch fullMsgHandledLatch = new CountDownLatch(1);
+
+    /** Number of the test nodes. */
+    private static final int NODES_CNT = 2;
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
+
+        if (getTestIgniteInstanceName(NODES_CNT - 1).equals(igniteInstanceName)) {
+            ((TestTcpDiscoverySpi) cfg.getDiscoverySpi()).discoveryHook(new DiscoveryHook() {
+                @Override public void beforeDiscovery(DiscoveryCustomMessage customMsg) {
+                    if (customMsg instanceof ServiceClusterDeploymentResultBatch) {
+                        fullMsgReceivedLatch.countDown();
+
+                        try {
+                            fullMsgUnblockedLatch.await(getTestTimeout(), MILLISECONDS);
+                        }
+                        catch (InterruptedException e) {
+                            throw new RuntimeException(e);
+                        }
+                    }
+                }
+
+                @Override public void afterDiscovery(DiscoveryCustomMessage customMsg) {
+                    if (customMsg instanceof ServiceClusterDeploymentResultBatch)
+                        fullMsgHandledLatch.countDown();
+                }
+            });
+
+            cfg.setUserAttributes(Collections.singletonMap(ATTR_SKIP_DEPLOYMENT, true));
+        }
+
+        return cfg;
+    }
+
+    /**
+     * Ignores the test in case the legacy service processor is used.
+     */
+    @Before
```

Comment:
`@BeforeClass` because all these tests are valid only for new service processor implementaion.